Added IAnvilRecipe

ShetiPhian opened the pull request:
Allows mods to define recipes for the anvil.

This is an update of my 1.8 version: https://github.com/MinecraftForge/MinecraftForge/pull/2389
And that was an update of DemoXinMC's pull request: https://github.com/MinecraftForge/MinecraftForge/pull/987

With the following changes:
1. minimum code in ContainerRepair, rest is moved into AnvilRecipes
2. added some comments (just some simple ones)
3. changed inputSlot1 and inputSlot2 to inputLeft & inputRight to correspond to the GUI and be similar to the existing anvil events.
4. renamed repair to doRepair, and removed the output stack as there is no point in editing it here because getResult is where you set it up.
5. added SimpleAnvilRecipe, (should cover most needs and supports the OreDictionary)
6. added getInputs() to IAnvilRecipe for use by recipe display mods
7. when renaming the new name is passed to the IAnvilRecipe

Includes a Test Mod:
-- 1x Leather Chest-plate + 5x Iron = 1x Iron Chest-plate
-- 1x Axe + 1x MelonBlock = 6x Melon Slices & -1 durability to axe
-- 1x LavaBucket + [1-64]x Slimeballs + [1-9]x Exp = [1-64]x MagmaCream & 1x EmptyBucket

jk-5 commented on src/main/java/net/minecraftforge/common/IAnvilRecipe.java:
```diff
@@ -0,0 +1,56 @@
+     * Returns the input for this recipe, any mod accessing this value should never
+     * manipulate the values in this array as it will effect the recipe itself.
+     */
+    List[] getInputs();
```
Missing generics

ShetiPhian replied:
It would be better for me to change this completely so it was more like the crafting table recipes, and return a ArrayList<Object> here instead.

Also noticed while I did replace all tabs with spaces I forgot to put the open brackets on new lines.

I'll get this fixed up.

darkhax commented:
While this can be done using AnvilUpdateEvent I like having this as its own thing.

liach commented on src/main/java/net/minecraftforge/common/AnvilRecipeHandler.java:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * Add an Anvil Recipe
+     */
+    public void addRepair(IAnvilRecipe recipe)
```
Just `add`, `remove`

liach commented on src/main/java/net/minecraftforge/common/IAnvilRecipe.java:
```diff
@@ -0,0 +1,64 @@
+     *
+     * @return The recipes input vales. Each value can be an ItemStack, List<ItemStack> or null.
+     */
+    ArrayList<Object> getInput();
```
I suggest to use `List<ItemStack> getInputLeft` and `List<ItemStack> getInputRight` methods, as `ArrayList` may not be a good return value, and single item stacks can be turned into a singleton list.

ShetiPhian replied:
I originally changed it to match the one in the crafting table recipes, but all that did was add complexity. I have a very similar recipe handler for one of my blocks, and after applying the same change to it I decided to change it back. Everything needed casting and checking before use, making it a complete pain.

diesieben07 commented on src/main/java/net/minecraftforge/common/AnvilRecipeHandler.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public void removeRepair(IAnvilRecipe recipe)
+    {
+        if (anvilRecipes.contains(recipe))
```
This check is pointless.

liach commented on src/main/java/net/minecraftforge/common/IAnvilRecipe.java:
```diff
@@ -0,0 +1,64 @@
+     * @param newName    The items new name, empty if not being renamed.
+     * @param world      The current world
+     */
+    ItemStack getResult(ItemStack inputLeft, ItemStack inputRight, String newName, World world);
```
Add boolean parameters for simulation?

liach commented on src/main/java/net/minecraftforge/common/SimpleAnvilRecipe.java:
```diff
@@ -0,0 +1,136 @@
+        return checkMatch(this.inputs.get(0), inputLeft, false) && checkMatch(this.inputs.get(1), inputRight, true);
+    }
+
+    private boolean checkMatch(Object target, ItemStack tableItem, boolean checksize)
```
IMO `Object target` can be changed to `List<ItemStack> target` and make life easier if you accept my suggestion.

liach commented on src/main/java/net/minecraftforge/common/AnvilRecipeHandler.java:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * Remove an Anvil Recipe
+     */
+    public void removeRepair(IAnvilRecipe recipe)
```
To return `boolean` type?

Actuarius commented:
@williewillus added labels [Feature]

ShetiPhian commented on src/main/java/net/minecraftforge/common/SimpleAnvilRecipe.java:
```diff
@@ -0,0 +1,142 @@
+        {
+            return tableItem == null;
+        }
+        else if (checksize && tableItem.stackSize < this.matCost)
```
Note: This line has a NPE if the material slot is empty.
Replace line with: `else if (checksize && (tableItem == null || tableItem.stackSize < this.matCost))`

ShetiPhian commented:
I don't believe I'm going to update this pull, personally I only have two recipes suited for anvil crafting and they can easily be changed to table recipes.

I still feel like anvil recipes are a good fit for Forge, so consider all the code up for grabs. Anyone who wants to can use it to make their own pull. (If this never happens and I do end up with extra free time I'll update, but that wont be for a while)

Notes for whomever decides to update:
1) _SimpleAnvilRecipe.checkMatch has a NPE_
Found this a while back, planned on fixing when I corrected merge conflicts.
I've added a line note with the code fix.

2) _You can not remove a renamed item from the anvil._
This was recently brought to my attention, I have not looked into it, but my guess is inverted logic.
